.github/workflows/main.yml

```diff
@@ -107,3 +107,20 @@ jobs:
             docker tag "$IMAGE_NAME" "$PUBLIC_IMAGE_NAME:latest"
             docker push "$PUBLIC_IMAGE_NAME:$IMAGE_VERSION"
             docker push "$PUBLIC_IMAGE_NAME:latest"
+
+  notify-slack-on-scheduled-failures:
+    needs: [build, test-docker-image, publish]
+
+    runs-on: ubuntu-latest
+
+    if: ${{ !cancelled() && github.event_name == 'schedule' && contains(needs.*.result, 'failure') }}
```

Comment:
I'm curious, why are you able to do `if: failure()` on the other workflow, but you need to do `contains(needs.*.result, 'failure')` here?

Reply:
@lucyb: Good question!

## Dev container test workflow

Let's look at the dev container test first, since it's simpler. It's a single workflow job.

We can use `if: failure()` there as a notification trigger as the [final step of that single job](https://docs.github.com/en/actions/learn-github-actions/expressions#failure)[^1]:

> Returns `true` when any previous step of a job fails.

So if any previous step of the job fails, a notification is sent.

## `main` workflow

Now on to the `main` workflow. There we have individual build, test and publish jobs.

We want to know if *any* of those three jobs fail in a scheduled workflow run.

To use `if: failure()` for that, we'd have to add the same check at the end of each individual job. You might be able to remove the repetitive duplication with a [reusable workflow](https://docs.github.com/en/actions/using-workflows/reusing-workflows#creating-a-reusable-workflow), but you'd still have an extra step added to each job.

Instead, it seemed tidier to check the result of all of these individual jobs in a single notification job. That's what `contains(needs.*.result, 'failure')` does.

Arguably, if we wanted to make the two more uniform, it might be fair to use this same structure of "separate notification job" in the dev container test workflow. But I'm not sure it's worth too much worrying about, because the dev container test workflow is simple anyway.

[^1]: It's incidental, but also maybe worth noting that [each workflow job step has a default implicit `if: success()` check if that is not overridden](https://docs.github.com/en/actions/learn-github-actions/expressions#status-check-functions).

Reply:
TIL! Thanks for that explanation. That's really clear and I've learned something new that might be useful for the future.